File: image-classification-py/train.py
###############################################
# Build an image classification model 
# with pre-trained network using Keras
###############################################

import os
import numpy as np
from keras.preprocessing.image import ImageDataGenerator
from keras.models import Sequential
from keras.layers import Dropout, Flatten, Dense
from keras import applications

# data directory
base_dir = "/home/dlvmadmin/.mlhub/image-classification-py"
train_data_dir = os.path.join(base_dir, "data/cats_and_dogs_small/train")
validation_data_dir = os.path.join(base_dir, "data/cats_and_dogs_small/validation")

# top model weights
bottleneck_features_train_path = os.path.join(base_dir, "bottleneck_features_train.npy")
bottleneck_features_validation_path = os.path.join(base_dir, "bottleneck_features_validation.npy")
top_model_path = os.path.join(base_dir, "bottleneck_fc_model.h5")
top_model_weights_path = os.path.join(base_dir, "bottleneck_fc_model_weights.h5")

# parameter values
img_width, img_height = 150, 150
nb_train_samples = 2000
nb_validation_samples = 1000
epochs = 30
batch_size = 20

# Define functions
def save_bottleneck_features():
    '''
    Create bottleneck features.
    '''
    # initiate the VGG16 network
    model = applications.VGG16(
        include_top=False, 
        weights='imagenet'
    )

    datagen = ImageDataGenerator(rescale=1. / 255)

    train_generator = datagen.flow_from_directory(
        train_data_dir,
        target_size=(img_width, img_height),
        batch_size=batch_size,
        class_mode=None,
        shuffle=False
    )

    bottleneck_features_train = model.predict_generator(
        train_generator, 
        steps=nb_train_samples // batch_size
    )

    np.save(open(bottleneck_features_train_path, 'wb'),
            bottleneck_features_train)

    validation_generator = datagen.flow_from_directory(
        validation_data_dir,
        target_size=(img_width, img_height),
        batch_size=batch_size,
        class_mode=None,
        shuffle=False
    )

    bottleneck_features_validation = model.predict_generator(
        validation_generator, 
        steps=nb_validation_samples // batch_size
    )

    np.save(open(bottleneck_features_validation_path, 'wb'),
            bottleneck_features_validation)


def train_top_model():
    '''
    Train top model.
    '''
    train_data = np.load(open(bottleneck_features_train_path, 'rb'))
    train_labels = np.array(
        [0] * int(nb_train_samples / 2) + [1] * int(nb_train_samples / 2))

    validation_data = np.load(open(bottleneck_features_validation_path, 'rb'))
    validation_labels = np.array(
        [0] * int(nb_validation_samples / 2) + [1] * int(nb_validation_samples / 2))

    model = Sequential()
    model.add(Flatten(input_shape=train_data.shape[1:]))
    model.add(Dense(256, activation='relu'))
    model.add(Dropout(0.5))
    model.add(Dense(1, activation='sigmoid'))

    model.compile(
        optimizer='rmsprop',
        loss='binary_crossentropy', 
        metrics=['accuracy']
    )

    model.fit(
        train_data, 
        train_labels,
        epochs=epochs,
        batch_size=batch_size,
        validation_data=(validation_data, 
                         validation_labels)
    )
    
    model.save(filepath=top_model_path, overwrite=True)
    model.save_weights(filepath=top_model_weights_path, overwrite=True)

# Create bottleneck features
save_bottleneck_features()

# Train top model
train_top_model()

File: dogs-cats-r/train.R
# load packages

library(httr)
set_config(config(ssl_verifypeer=0L))

#library(tfruns)
#library(reticulate)
#use_virtualenv("r-tensorflow", required = TRUE)

library(keras)

# data directory

base_dir <- "~/.mlhub/dogs-cats-r-r/data/cats_and_dogs_small"
train_dir <- file.path(base_dir, "train")
validation_dir <- file.path(base_dir, "validation")

# instantiate the VGG16 model

conv_base <- application_vgg16(
  weights = "imagenet",
  include_top = FALSE,
  input_shape = as.integer(c(150, 150, 3))
)

summary(conv_base)

# create keras sequential model by adding base model and layers

model <- keras_model_sequential() %>% 
  conv_base %>% 
  layer_flatten() %>% 
  layer_dense(units = 256, activation = "relu") %>% 
  layer_dense(units = 1, activation = "sigmoid")

summary(model)

length(model$trainable_weights)

freeze_weights(conv_base)
length(model$trainable_weights)

# data augmentation on training data

train_datagen = image_data_generator(
  rescale = 1/255,
  rotation_range = 40,
  width_shift_range = 0.2,
  height_shift_range = 0.2,
  shear_range = 0.2,
  zoom_range = 0.2,
  horizontal_flip = TRUE,
  fill_mode = "nearest"
)

# note that the validation data shouldn't be augmented

test_datagen <- image_data_generator(rescale = 1/255)  

# train model using the image data generator

train_generator <- flow_images_from_directory(
  train_dir,                  # Target directory  
  train_datagen,              # Data generator
  target_size = c(150, 150),  # Resizes all images to 150 × 150
  batch_size = 20,
  class_mode = "binary"       # binary_crossentropy loss for binary labels
)

validation_generator <- flow_images_from_directory(
  validation_dir,
  test_datagen,
  target_size = c(150, 150),
  batch_size = 20,
  class_mode = "binary"
)

model %>% compile(
  loss = "binary_crossentropy",
  optimizer = optimizer_rmsprop(lr = 2e-5),
  metrics = c("accuracy")
)

history <- model %>% fit_generator(
  train_generator,
  steps_per_epoch = 100,
  epochs = 30,
  validation_data = validation_generator,
  validation_steps = 50
)

# plot training history

plot(history)

cat("====================\nModel Saved as HDF5\n====================\n\n")

save_model_hdf5(model, 
                filepath = "dogs-cats-r-vgg16-model.hdf5", 
                overwrite = TRUE,
                include_optimizer = FALSE)

# Q: how to save model with training configuration?

save_model_weights_hdf5(model,
                        filepath = "dogs-cats-r-vgg16-model-weights.hdf5",
                        overwrite=TRUE)

cat("====================\nHistory Saved as RData and PNG\n====================\n\n")

save(history, file="dogs-cats-r-vgg16-history.RData")

png(filename = "history.png")
plot(history)
dev.off()
